### ISSUE TITLE
Connector Metadata Service

### ISSUE THREAD

**Issue by bnchrch** [airbytehq/airbyte#24011]
Tech spec: https://docs.google.com/document/d/1CXG0hTftomCTO3vB0ocbs00CkYBD6YURfrxm8KwozB4/edit#

**Comment by evantahler** [comment]
Nested epics - love it.

**Comment by bnchrch** [comment]
Solicited Metadata Project Update: Taking Shape! :sparkles:

Im back with another (now solicited) update as we continue towards getting this metadata project out the door! :rocket:
TL;DR: We've made significant progress on catalog generation and integrating with Dagger, with a few challenges along the way. :muscle:

:white_check_mark: What's been completed:
1. We can now override individual catalog definitions.
2. We can now generate the catalog from metadata files.
3. We can now ingest data from GitHub and spec cache.
4. We have some QA reports to catch any diffs between the old catalog and the new catalog (there should be near none).
5. We've integrated the metadata service into a **DAGGER** CI pipeline and updated the Dagger implementation along the way.
6. We've reached a decision on airbyte-types; this won't be a separate repo but stay tuned for `metadata_service/types`.

:construction: What got in the way:
1. The airbyte-types discussion became larger than expected.
2. Dagger's CI pipelines have a learning curve.
3. Dagger is still containers, and debugging dependencies is still hard!
4. Dagster types required some hand-holding to work with Pandas.
5. PR reviews took a little longer than expected.

:question: Learning / Questions / Changes:
Nothing in particular, looking forward to knowledge sharing in Q2!

:seedling: What's next:
1. Wrapping up the new Dagger pipelines.
2. Validating metadata.yml files before merge.
3. Uploading metadata.yml files on merge.

:no_entry: Blockers:
None at this time!

As always, your feedback and suggestions are much appreciated!

**Comment by bnchrch** [comment]
Metadata Project Update: Cloud Deoployed, Metadata Files Added
TL;DR: Dagger and Dagster are now officially friends, metadata files actually exist and Dagster is not just that weird thing you run locally, its now that weird thing you deploy to k8s.
:white_check_mark: What's been completed:
We have a PR that deploys metadata service jobs to Dagster Cloud automatically! :confetti_ball:
Dagger now runs our metadata library and orchestrator unit tests.
Dagger validates metadata files in your PRs!
Augustin put up a massive PR for adding all the metadata files to each connector! :clap: This should be merged Today.
We set up Slack alerts for Dagster.
We have a draft PR that adds the Dagster sensor for monitoring changes to metadata files in GCS.
A bunch of QoL improvements have been made to Dagger and the CLI.
:books: Learnings:
Dagster hybrid cloud required some unexpected Dockerfile and GitHub action magic to get working.
etags are confirmed awesome, it feels really neat knowing if files have changed without downloading them.
:construction: Hurdles going into this week:
Tim had a fair amount of Terraform and K8S work that needs untangling. The Prod Eng team is helping with this.
Dagster has a new environment variable story that is great but requires updating the Dagster agent in our K8S cluster. The Prod Eng team is assisting here as well.
:seedling: What's next:
Merging the massive PR for adding all the metadata files to each connector.
Finalizing the Dagster sensor for monitoring changes to metadata files in GCS.
Work with prodeng to update the Dagster Agent to latest
Add the job that actually parses these metadata files into the final catalog
:no_entry_sign: Blockers:
Dagster cloud is blocked from prime time until we update the Dagster Agent in the helm chart. A pre-requisite here seems to be getting tailscale gating access to the dagster cluster.

**Comment by bnchrch** [comment]
*Metadata Project Update: Hosted Dagster, Spec Mask Assets, and Sensors!*

**:white_check_mark: What's been completed:**
1. Updated Dagster to deploy to the hosted version of Dagster Cloud.
2. Metadata files are now in master (and the dev team has been made aware).
3. Metadata file GCS sensor is merged.
4. PR up for renaming the Catalog to Registry.
5. Confirmed a plan with Prod Eng to use our load balancer to help us migrate from the old catalogs to the new registry.
6. Draft PR up for generating the final catalog from the metadata sensor and ensuring it conforms to the ConnectorRegistry JSON schema. (@Pedro Lopez (Airbyte) Thanks for getting this type into platform)
7. PR ready to be merged for a new Dagster job that generates the spec mask. (Thanks @Augustin Lafanechere (Airbyte))

**:books: Learnings and hurdles:**
1. Dagster Cloud Serverless had a hard time with relative imports/poetry, which took additional effort to resolve.
2. Dagger pipelines have been slow. Dagger is working on a fix, but it has definitely impacted the speed at which we merge.

**:no_entry_sign: Blockers:**
1. Dagger pipelines are currently slow, delaying PR merges until resolved.
2. Next week depends on ProdEng work, which is scheduled for Wednesday.
3. I'm on call this week. :disappointed: 

**:seedling: Looking ahead:**
1. Expect the new catalog generation to be ready this week and deployed along side our current catalog at the top of next week.
2. Creating a successor to the connectors.md and builds.md files generated by Dagster.
3. New registry paths available in connectors.airbyte.com, making a deployment of the new catalog ready for next week.
4. The word "catalog" disappearing.

**Comment by bnchrch** [comment]
*Metadata Project Update: Registry Progress, New Report, and Deploy Fixes!*

✅ What's been accomplished:
1. Merged in the naming update from Catalog to Registry.
2. Merged in job that generates the new Registry on metadata file changes.
3. Merged in a [new report](https://storage.googleapis.com/ben-ab-test-bucket/generated_reports/connector_registry_report.html) that will replace `builds.md` and `connectors.md`.
4. Tested and deployed new URL paths for the load balancer in front of https://connectors.airbyte.com so that we can easily switch between the new catalog and the old catalog with minimal change.
5. Two PRs up for changing airbyte and platform to point at the new registry URLs.
6. A PR up to introduce mandatory connector PR checklists. This will help ensure people know that the definition files are gone when we launch.

📚 Learnings and hurdles:
1. On-call took some time away last week and this week.
2. ProdEng is a delight to work with.

🚫 Blockers:
1. There is a deploy problem with Dagster Cloud to investigate today.
2. I need to update some Dagger pipelines I removed Docker login from.

🌱 Looking ahead:
1. The next steps are all focused on releasing the new registries, so we will be:
   1. Updating Docs.
   2. Pointing the new registry systems at the production bucket.
   3. Releasing the new report to be used by the rest of the company.
   4. Updating the registry URLs in our repos.
   5. Updating the metadata files to ensure a near-zero diff between the old and new registry before going live.

**Comment by sw-yx** [comment]
linking some previous work we did: https://github.com/airbytehq/airbyte/issues/17893

and the data that resulted https://airbyte-metadata.netlify.app/

**Comment by bnchrch** [comment]
*Metadata Project Update: Registry Progress, New Report, and Deploy Fixes!*

✅ What's been accomplished:
1. Our repos are now all successfully pointing at https://connectors.airbyte.com/files/registries/v0/oss_registry.json 🎉
2. Released the new Connector Registry Report
3. Metadata diff at zero again
4. feat branch: Removed all references in Python files, and shell scripts to the `_definition.yaml` files
5. PR up to remove all references in docs and java to `_definition.yaml` files in `airbyte`
6. Draft PR up to remove all references to `_definition.yaml` files in `platform-internal`

📚 Learnings and hurdles:
1. Our team is awesome online AND in person
2. CI really made new URL take a while

🚫 Blockers:
Unfortunately the new dagger pipelines had some issues yesterday preventing metadata from auto uploading. Will be triaging today but it definately will push our roll out of publish on merge.

May end up releasing the new connector registry outside of publish on merge.

🌱 Looking ahead:
1. Cut over to the new metadata registry!
2. If successful merge in the feature branch for removing the legacy yaml files
3. Add a few straggling checks like if the normalization docker image exists

**Comment by bnchrch** [comment]
*Metadata Project Update: One integration test away*

**✅ What's been accomplished:**
1. Successfully redirected the load balancer to the new registry files.
2. Removed actor definition files from `airbyte`, reducing 60,000 lines of code.
3. Deployed a backup plan to upload new metadata files via GitHub Actions.
4. Updated the Dagster sensors to use MD5 instead of ETags.
5. Updated `airbyte` and `platform-internal` to use the remote spec secret mask.
6. Made a number of small fixes along the way.

**🚀 In Progress:**
1. Checking that normalization images exist.
2. Removing actor definition files from `platform-internal`.
3. Hackday project: Metadata Icon CDN.

**⚠️ What got in the way:**
1. A P1 incident related to a bad spec and GCS caching.
2. Hackdays.
3. Failing integration tests in the dev environment.
4. A few forgetful incidents where `source-file-secure` was not updated.
5. Regular CI and dev environment issues.

**🤝 What I could use a hand with:**
1. Running cloud integration tests locally. (Wanted to wait until after hackdays before pinging anyone)

**🔮 What's next:**
1. Fix the integration test for the platform's actor definitions removal and merge it!
2. Add tags to metadata type.
3. Polish up the Icon CDN PR.

**Comment by bnchrch** [comment]
*Metadata Project Update: Icons and Tags Live 🌟*

**:white_check_mark: What's been accomplished:**
1. `iconUrls` are now live in the Registry.
2. Added the `tags` field to the metadata and registry. (Every connector now must have at least one `language:<LANG>` tag.)
3. We now ensure that the spec is always uploaded.
4. Successfully removed a few thousand more lines of code related to the development of the new registry system.

**:rocket: In Progress:**
1. The QA engine is in the process of being moved to `airbyte-ci` and being updated to use `poetry`.

**:warning: What got in the way:**
A number of small bugs and issues with metadata and the pipeline have surfaced:
1. Bug: The spec was not being uploaded on prepublish.
2. Bug: An invalid Metadata extension was causing a skipped check.
3. Issue: A connector had two docker image versions set in its Dockerfile.
4. Bug: Icons were not displaying in the connector reports.
5. Task: `source-amazon-seller` needed to be removed from the cloud.

**:construction: Blockers:**
Need access to a production bucket to confirm that the QA-engine is indeed working as expected. Currently Alex G is helping me with that

**:crystal_ball: What's next:**
1. QA Engine fixes and migration.
2. Updating our docs to include a guide on how to generate a catalog.
3. Updating nightlies to use the new pipeline.
4. Updating Dagster to output the nightly report.
5. Adding Destination Types to the registry.
6. Utilizing IconUrls in the platform.

**Note:** As we're moving towards the completion of the Metadata project, updates on this project will become less frequent and smaller in scale. The focus will shift towards optimizing and maintaining the current system.

**Comment by bnchrch** [comment]
*Metadata Project Update: Towards a Smoother CI/CD Pipeline* 🚀

**✅ What's been accomplished:**
1. The QA Engine has been moved to `airbyte-ci` and now utilizes poetry for deterministic installs.
2. We've updated our 'Generate Catalog locally' docs to mirror the new Registry/Dagster view.
3. Nightlies (and build, test, publish) now output their results to `airbyte-ci-reports` in GCS.
4. Added a new Dagster sensor to watch GitHub Actions and trigger asset builds upon completion.
5. Upgraded our docker image existence check using the registry API to fix issues with false negatives when DockerHub goes down.
6. We've improved Metadata Validation to catch new connectors without a metadata file.

**🚀 In Progress:**
1. We're revamping the Slack connector health report and badges to use the new logs in `airbyte-ci-reports`.
2. Working on consolidating all CI service accounts into one for the CI pipeline.

**⚠️ What got in the way:**
1. DockerHub API went down.
2. Technical spec discussions took some time.
3. CAT test issues.
4. Issues with Registry being re-added to GitHub.
5. Some PRs skipping the metadata validation.

**🔮 What's next:**
1. We'll introduce a new role in GCP that denotes CI access.
2. Updating Dagster to output the nightly report.
3. Adding Destination Types to the registry.
4. Using IconUrls in the platform.

**Note:** As we're nearing the finish line for the Metadata project, updates will become less frequent and smaller. We're shifting our focus to optimization and maintenance of the existing system.
